src/runtime/js/ts.ui/core/core-api@tradeshift.com/models/dialogs/ts.ui.SpotlightModel.js

```diff
@@ -0,0 +1,188 @@
+/**
+ * Advanced dialog model.
+ * @extends {ts.ui.Model}
+ * @using {ts.ui.Dialog} Dialog
+ * @using {gui.Combo#chained} chained
+ * @using {gui.Type} Type
+ * @using {gui.Object} GuiObject
+ * @using {ts.ui.String} Constants
+ * @using {string} primary
+ * @using {string} secondary
+ * @using {string} tertiary
+ */
+ts.ui.SpotlightModel = (function using(CoverSpirit, chained, Type, GuiObject, Constants, primary, secondary, tertiary) {
+    /**
+     * Get JSON for button where the properties of the passed
+     * JSON will overwrite the given (default) label and type.
+     * The {ts.ui.DialogSpirit} may modify the `type` property.
+     * @param {object} json Button configuration
+     * @param {string} id So that we can find it later
+     * @param {string} label Default label
+     * @param {string} type Default type
+     * @returns {object}
+     */
+
+	return ts.ui.Model.extend({
+
+		steps: [],
+
+		onconstruct: function() {
+			this.super.onconstruct();
+			this.steps = this.steps || [];
+		},
+
+		isOffScreen: function(content) {
+			var rect = content.getBoundingClientRect();
+			return (rect.right + rect.width || rect.top + rect.height) < 0 ||
+                (rect.right > window.innerWidth || rect.top > window.innerHeight);
+		},
+
+		open: function() {
+			var titleText = this.step.title;
+			var hintText = this.step.hint;
+			var radius = this.step.radius;
+			var targetElement = this.step.element;
+
+			targetElement.scrollIntoView({
+				behavior: 'smooth',
+				block: 'end'
+			});
+
+			var targetPos;
+			if (targetElement) {
+				targetPos = this.getPosition(targetElement);
+			}
+
+			var cover = ts.ui.CoverSpirit.getCover('ts-spotlightcover');
+			var existingContent = document.querySelectorAll('.ts-spotlight-content');
+			var existingButton = document.querySelectorAll('.ts-spotlight-close-button');
+
+			if (existingContent.length) {
+				existingContent[0].remove();
+			}
+
+			if (existingButton.length) {
+				existingButton[0].remove();
+			}
+
+			var content = document.createElement('div');
+			content.style.position = 'absolute';
+			content.style.color = 'white';
+			content.className = 'ts-spotlight-content ts-transition';
+
+			content.style.left = targetPos.x + radius + 22 + 'px';
+			content.style.top = targetPos.y + (radius / 6) + 'px';
+
+			content.style.fontWeight = 'normal';
+			content.style.paddingRight = '10px';
+			content.style.zIndex = '3999';
+			content.style.width = '300px';
+
```

Comment:
Perhaps some of these styles should go into CSS to cut down on the code?

It's by the way possible to mass-declare style properties using proprietary technology.
```javascript
var content: gui.CSSPlugin.style(document.createElement('div', {
	position: 'absolute',
	color: 'white',
	left: targetPos.x + radius + 22 + 'px',
	top: targetPos.y + (radius / 6) + 'px',
	fontWeight: 'normal',
	paddingRight: '10px',
	zIndex: '3999',
	width: '300px'
});
```